vpr/src/place/net_cost_handler.cpp

```diff
@@ -0,0 +1,2277 @@
+
+/* [0...cluster_ctx.clb_nlist.nets().size()-1]                                               *
+ * A flag array to indicate whether the specific bounding box has been updated   *
+ * in this particular swap or not. If it has been updated before, the code       *
+ * must use the updated data, instead of the out-of-date data passed into the    *
+ * subroutine, particularly used in try_swap(). The value NOT_UPDATED_YET        *
+ * indicates that the net has not been updated before, UPDATED_ONCE indicated    *
+ * that the net has been updated once, if it is going to be updated again, the   *
+ * values from the previous update must be used. GOT_FROM_SCRATCH is only        *
+ * applicable for nets larger than SMALL_NETS and it indicates that the          *
+ * particular bounding box cannot be updated incrementally before, hence the     *
+ * bounding box is got from scratch, so the bounding box would definitely be     *
+ * right, DO NOT update again.                                                   */
+static vtr::vector<ClusterNetId, NetUpdateState> bb_updated_before;
+
+/* The following arrays are used by the try_swap function for speed.   */
+
+/**
+ * The wire length estimation is based on the bounding box of the net. In the case of the 2D architecture,
+ * we use a 3D BB with the z-dimension (layer) set to 1. In the case of 3D architecture, there 2 types of bounding box:
+ * 3D and per-layer. The type is determined at the beginning of the placement and stored in the placement context.
+ *
+ *
+ * If the bonding box is of the type 3D, ts_bb_coord_new and ts_bb_edge_new are used. Otherwise, layer_ts_bb_edge_new and
+ * layer_ts_bb_coord_new are used.
+ */
+
+/* [0...cluster_ctx.clb_nlist.nets().size()-1] -> 3D bounding box*/
+static vtr::vector<ClusterNetId, t_bb> ts_bb_coord_new, ts_bb_edge_new;
+/* [0...cluster_ctx.clb_nlist.nets().size()-1][0...num_layers] -> 2D bonding box on a layer*/
+static vtr::vector<ClusterNetId, std::vector<t_2D_bb>> layer_ts_bb_edge_new, layer_ts_bb_coord_new;
+/* [0...cluster_ctx.clb_nlist.nets().size()-1][0...num_layers] -> number of sink pins on a layer*/
+static vtr::Matrix<int> ts_layer_sink_pin_count;
+/* [0...num_afftected_nets] -> net_id of the affected nets */
+static std::vector<ClusterNetId> ts_nets_to_update;
+
+/**
+ * @param net
+ * @param moved_blocks
+ * @return True if the driver block of the net is among the moving blocks
+ */
+static bool driven_by_moved_block(const ClusterNetId net,
+                                  const int num_blocks,
+                                  const std::vector<t_pl_moved_block>& moved_blocks);
+/**
+ * @brief Update the bounding box (3D) of the net connected to blk_pin. The old and new locations of the pin are
+ * stored in pl_moved_block. The updated bounding box will be stored in ts data structures.
+ * @param net
+ * @param blk
+ * @param blk_pin
+ * @param pl_moved_block
+ */
+static void update_net_bb(const ClusterNetId& net,
```

Comment:
Given that this function is also used for 2D architectures, I decided not to change its name to avoid confusion. For the other function (update_net_layer_bb), I included "layer" in the name because it is only used for 3D architecture. If you think it is okay to change the name of this function to update_net_3Dbb, please let me know.

Reply:
OK, in that case just make sure there is a nice comment on the cubeBB and layer_BB explaining them both in detail near the top of the file.  That comment can say cubeBB is a 3D bounding box which degenerates to the usual 2D BB for one layer devices.